### ISSUE TITLE
(master) links shows error when restarting linked-to container

### ISSUE THREAD

**Issue by cpuguy83** [docker/docker#12334]
With master binary:

```
# docker run -d --name test busybox top
54ae2450e1fcccbb53552b904cabef462718fe12da271a772116374be7a9eb1a
# docker run -d --name test2 --link test:test busybox
cd5ef2344c99c5fa8fc8ad49c9e1d63a4966b489ff3672f05185cbb2c411ff06
# docker rm test2
# docker restart test
DEBU[0096] Calling POST /containers/{name:.*}/restart
INFO[0096] POST /v1.19/containers/test/restart?t=10
INFO[0096] +job restart(test)
DEBU[0096] Sending 15 to 54ae2450e1fcccbb53552b904cabef462718fe12da271a772116374be7a9eb1a
ERRO[0096] no such id: cd5ef2344c99c5fa8fc8ad49c9e1d63a4966b489ff3672f05185cbb2c411ff06
INFO[0097] -job restart(test) OK
test
#
```

Links still work, however it seems removed containers are not being de-refed properly in the link graph.

**Comment by runcom** [comment]
I'll try to  dig arund

**Comment by runcom** [comment]
#dibs

**Comment by calavera** [comment]
Closing as fixed by #12369.
